tests/unit/common/eeprom/eeprom_acces.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+#include <vector>
+#include <stdint.h>
+#include "crc32.h"
+#include <stdlib.h>
+#include <string.h>
+#include "msgpack.h"
+#include <optional>
+
+// helper structs
+struct Key {
+    uint32_t key;
+    template <class Packer>
+    void pack(Packer &pack) {
+        pack(key);
+    }
+};
+
+template <class T>
+struct DataItem {
+    uint32_t key;
+    T data;
+
+    DataItem(uint32_t key, T data)
+        : key(key)
+        , data(data) {}
+    DataItem() = default;
+
+    template <class Packer>
+    void pack(Packer &pack) {
+        pack(key, data);
+    }
+};
+
+class EepromAccess {
+#ifdef EEPROM_UNITTEST
+public:
+#endif
+    static constexpr uint16_t START_OFFSET = 0x500;
+    static constexpr uint8_t CUSTOM_ITEM_ID = 0x7f;
+    static constexpr uint8_t MSGPACK_EXT8 = 0xc7;
+    static constexpr uint8_t LAST_ITEM_STOP = 0xfe;
+    static constexpr size_t HEADER_SIZE = 3;
+    static constexpr size_t CRC_SIZE = 4;
+
+    std::vector<std::pair<uint32_t, uint16_t>> index;
+
+    uint16_t first_free_space = START_OFFSET;
+    uint16_t write_stop = 8096; //<FIXME should be set as part of init proccess
+
+    template <class T>
+    std::vector<uint8_t> static serialize_data(const char *key, T data);
+    template <class T>
+    DataItem<T> static deserialize_data(std::vector<uint8_t> data);
+
+    // find address for hash in index
+    std::optional<uint16_t> get_address(uint32_t hash);
+
+    // stores item to eeprom and calculates the crc of the data
+    // creates header of the data
+    // the header  element ext 8 from msgpack spec and the value for type is EepromAccess::CUSTOM_ITEM_ID
+    // +--------+--------+--------+========+
+    // |  0xc7  |XXXXXXXX|  type  |  data  |
+    // +--------+--------+--------+========+
+    // where XXXXXXXX is a 8-bit unsigned integer which represents the length of the data including crc
+    void store_item(const std::vector<uint8_t> &data);
+
+    // load item from eeprom, if item is invalid it returns nullopt
+    std::optional<std::vector<uint8_t>> load_item(uint16_t address);
+
+public:
+    // iterate through eeprom and check it, calc crc for every item and fills the index
+    // load data to the eeprom struct?
+    void Init();
+
+    // remove duplicates of stored data
+    void Cleanup();
+
+    // Saves data item after last item
+    // if it hits max number of items it will remove duplicate items
+    template <class T>
+    void Set(const char *key, const T &data);
+
+    // looks in the index if we have the key, if yes it load s the data from eeprom if not returns nullopt
+    // should be called only after init before any call to set, because set can invalidate the index
+    template <class T>
+    std::optional<T> Get(const char *key);
```

Comment:
* It would be really nice if the type was checked at compile time or at least by some kind of assert. I mean, the code should be 100% sure what type it is. Or do I get it wrong when it can give me `nullopt`?
* I don't like the stuff about invalidation. The eeprom is being used in a way where someone is setting it and someone (potentially someone _else_) is reading it without the knowledge of the set.

Reply:
- This should give you nullopt when the data are not in eeprom or are corrupted. 
- You could get bad data if you use different type than the one which is in eeprom. 
- Eeprom access does not know which data type is stored in eeprom at compile time. 
- Eeprom access shoul be used just through Eeprom, which knows how to use it.